### ISSUE TITLE
Suggested improvements to the `Trial` class 

### ISSUE THREAD

**Issue by lebrice** [Epistimio/orion#872]
Here are some of my thoughts about the `Trial` class:

- Creating `Trial` objects is too cumbersome atm in my opinion.
    
    To create a `Trial`, you can either pass a  invariably have to go through one of the factory functions from `format_trials`:
    ```python3
    from orion.core.utils import format_trials
    space = SpaceBuilder().build({"x": "uniform(0, 1)", "y": "uniform(0, 1)"})  # Already a bit much imo
    trial = format_trials.dict_to_trial({"x": 0.1, "y": 0.2}, space=space)
    ```
    
    Since we always have to have the space to create a `Trial` anyway, then why not do something like this instead?
    ```python
    trial = Trial({"x": 0.1, "y": 0.2}, space=space)
    ```

- What was the original reasoning behind using the `Trial.Param` objects in `_params`, vs just dicts of the parameters themselves?
    - If the answer has to do with the types / bounds / etc, then that information is already in the Space, no need to duplicate it in the Trial as well imo. Adding a space attribute could help with this.
- `Trial` deserves a fully-specified set of (type annotated) constructor arguments.
    - There's no real way of knowing what is expected to be passed atm without digging through the code.

- `Trial` should act like a dictionary of parameter names to parameter values.
    - Current approach of _having_ a dictionary, rather than _being_ one isn't that natural IMO, especially when combined with spaces.
    
    For example, the following would look quite natural:.
    ```python
    space = SpaceBuilder().build({"x": "uniform(0, 1)", "y": "uniform(0, 1)"})
    assert 0.0 <= space.sample()["x"] < 1.0 
    ```

- The `Trial.objective` should be settable with a `float` value, in which case it just creates whatever kind of `Trial.Result` it needs, and sets its value to the given value.

**Comment by bouthilx** [comment]
I agree that would be much simpler. The `Params` object was a sql-like way of representing the data in the database which never proved useful and will probably never will. I would be happy to get rid of it as it always made the handling of `_params` quite cumbersome. 

Trial being a dictionary is a focus on the hyperparameters solely. The trials are more than that as the represent the status of execution and results. I think the dictionary interpretation of Trial you have in mind could be called a `Point` or `Configuration`. That's basically current `trial.params`.

The objective has been hard-coded as the sole objective of the trial but we should depart from this assumption. We are already starting having issues with the integration of Ax's multi-objective optimization because of that. It could be nice however to change the objective/stats/constraints interface so that we can set it more easily like `trial.objectives[some_name] = some_value` or `trial.set_objective(some_name, some_value)`.
